Exception when trying to save linked file

<!-- Note: Please use the GitHub Issue tracker only for BugReports. 
Feature requests,  questions and general feedback is now handled at http://discourse.jabref.org 
Thanks! --> 

- [X ] I have tested the latest master version from http://builds.jabref.org/master/ and the problem persists

JabRef 4.2-dev--snapshot--2018-03-06--master--bb471a26d
Linux 4.4.0-104-generic amd64 
Java 1.8.0_161


Steps to reproduce:

1. Create an entry with a URL leading to a PDF file, like this: https://www.gla.ac.uk/media/media_452489_en.pdf
2. Right-click the entry in the entry table.
3. Choose "Copy linked files to folder...".
4. Choose your home folder.

<!-- If applicable, excerpt of the bibliography file, screenshot, and excerpt of log (available in the error console) -->

<details>
File: copyfilesdialog.fxml not found, attempting with camel case
10:16:05.694 [JavaFX Application Thread] ERROR org.jabref.FallbackExceptionHandler - Uncaught exception occurred in Thread[JavaFX Application Thread,5,main]
java.lang.IllegalStateException: Cannot load org.jabref.gui.copyfiles.copyfilesdialog
	at com.airhacks.afterburner.views.FXMLView.loadSynchronously(FXMLView.java:101) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	at com.airhacks.afterburner.views.FXMLView.initializeFXMLLoader(FXMLView.java:125) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	at com.airhacks.afterburner.views.FXMLView.getView(FXMLView.java:137) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	at org.jabref.gui.AbstractView.getView(AbstractView.java:28) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	at org.jabref.gui.copyfiles.CopyFilesDialogView.show(CopyFilesDialogView.java:25) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	at org.jabref.gui.copyfiles.CopyFilesAction.showDialog(CopyFilesAction.java:63) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	at org.jabref.gui.copyfiles.CopyFilesAction.lambda$startServiceAndshowProgessDialog$3(CopyFilesAction.java:58) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	at com.sun.javafx.application.PlatformImpl.lambda$null$172(PlatformImpl.java:295) ~[jfxrt.jar:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_161]
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$173(PlatformImpl.java:294) ~[jfxrt.jar:?]
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95) ~[jfxrt.jar:?]
	at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method) ~[jfxrt.jar:?]
	at com.sun.glass.ui.gtk.GtkApplication.lambda$null$48(GtkApplication.java:139) ~[jfxrt.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_161]
Caused by: javafx.fxml.LoadException: 
file:/home/erelsgl/Dropbox/papers/JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar!/org/jabref/gui/copyfiles/CopyFilesDialog.fxml

	at javafx.fxml.FXMLLoader.constructLoadException(FXMLLoader.java:2601) ~[jfxrt.jar:?]
	at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2571) ~[jfxrt.jar:?]
	at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2441) ~[jfxrt.jar:?]
	at javafx.fxml.FXMLLoader.load(FXMLLoader.java:2409) ~[jfxrt.jar:?]
	at com.airhacks.afterburner.views.FXMLView.loadSynchronously(FXMLView.java:99) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	... 13 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:71) ~[?:1.8.0_161]
	at sun.reflect.GeneratedMethodAccessor6.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:275) ~[?:1.8.0_161]
	at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2566) ~[jfxrt.jar:?]
	at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2441) ~[jfxrt.jar:?]
	at javafx.fxml.FXMLLoader.load(FXMLLoader.java:2409) ~[jfxrt.jar:?]
	at com.airhacks.afterburner.views.FXMLView.loadSynchronously(FXMLView.java:99) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	... 13 more
Caused by: java.lang.NullPointerException
	at java.util.AbstractCollection.addAll(AbstractCollection.java:343) ~[?:1.8.0_161]
	at javafx.collections.ModifiableObservableListBase.addAll(ModifiableObservableListBase.java:99) ~[jfxrt.jar:?]
	at javafx.beans.binding.ListExpression.addAll(ListExpression.java:266) ~[jfxrt.jar:?]
	at org.jabref.gui.copyfiles.CopyFilesDialogViewModel.<init>(CopyFilesDialogViewModel.java:14) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	at org.jabref.gui.copyfiles.CopyFilesDialogController.initialize(CopyFilesDialogController.java:34) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:71) ~[?:1.8.0_161]
	at sun.reflect.GeneratedMethodAccessor6.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:275) ~[?:1.8.0_161]
	at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2566) ~[jfxrt.jar:?]
	at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2441) ~[jfxrt.jar:?]
	at javafx.fxml.FXMLLoader.load(FXMLLoader.java:2409) ~[jfxrt.jar:?]
	at com.airhacks.afterburner.views.FXMLView.loadSynchronously(FXMLView.java:99) ~[JabRef-4.2-dev--snapshot--2018-03-06--master--bb471a26d.jar:?]
	... 13 more

</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -51,6 +51,7 @@ The new default removes the linked file from the entry instead of deleting the f
 - We fixed an issue where not all bibtex/biblatex fields would be exported as latex-free to MS-Office XML [koppor#284](https://github.com/koppor/jabref/issues/284)
 - We fixed an issue where linked files would be deleted from bibliography entries despite choosing the "Cancel" option in the dialog menu.
 - We fixed the name of the group editing window to "Add group" instead of "Edit Group" when adding a new group. [koppor#277](https://github.com/koppor/jabref/issues/277)
+- We fixed an issue where the "Copy linked files" dialog produced an error when the entry had no file [#3818](https://github.com/JabRef/jabref/issues/3818)
 
 ### Removed
 - We removed the [Look and Feels from JGoodies](http://www.jgoodies.com/freeware/libraries/looks/), because the open source version is not compatible with Java 9.

--- a/src/main/java/org/jabref/gui/copyfiles/CopyFilesAction.java
+++ b/src/main/java/org/jabref/gui/copyfiles/CopyFilesAction.java
@@ -53,7 +53,6 @@ public class CopyFilesAction extends AbstractAction {
     private void startServiceAndshowProgessDialog(Task<List<CopyFilesResultItemViewModel>> exportService) {
 
         DefaultTaskExecutor.runInJavaFXThread(() -> {
-
             dialogService.showCanceableProgressDialogAndWait(exportService);
         });
 
@@ -64,8 +63,6 @@ public class CopyFilesAction extends AbstractAction {
     }
 
     private void showDialog(List<CopyFilesResultItemViewModel> data) {
-
-
         CopyFilesDialogView dlg = new CopyFilesDialogView(databaseContext, new CopyFilesResultListDependency(data));
         dlg.show();
     }

--- a/src/main/java/org/jabref/gui/copyfiles/CopyFilesAction.java
+++ b/src/main/java/org/jabref/gui/copyfiles/CopyFilesAction.java
@@ -52,13 +52,20 @@ public class CopyFilesAction extends AbstractAction {
 
     private void startServiceAndshowProgessDialog(Task<List<CopyFilesResultItemViewModel>> exportService) {
 
-        DefaultTaskExecutor.runInJavaFXThread(() -> dialogService.showCanceableProgressDialogAndWait(exportService));
+        DefaultTaskExecutor.runInJavaFXThread(() -> {
 
-        exportService.run(); //Run kinda blocks, so we just show the result dialog wgeb run is ready
-        DefaultTaskExecutor.runInJavaFXThread(() -> showDialog(exportService.getValue()));
+            dialogService.showCanceableProgressDialogAndWait(exportService);
+        });
+
+        exportService.run();
+        DefaultTaskExecutor.runInJavaFXThread(() -> {
+            showDialog(exportService.getValue());
+        });
     }
 
     private void showDialog(List<CopyFilesResultItemViewModel> data) {
+
+
         CopyFilesDialogView dlg = new CopyFilesDialogView(databaseContext, new CopyFilesResultListDependency(data));
         dlg.show();
     }

--- a/src/main/java/org/jabref/gui/copyfiles/CopyFilesTask.java
+++ b/src/main/java/org/jabref/gui/copyfiles/CopyFilesTask.java
@@ -38,7 +38,7 @@ public class CopyFilesTask extends Task<List<CopyFilesResultItemViewModel>> {
     private final long totalFilesCount;
     private final List<BibEntry> entries;
     private final List<CopyFilesResultItemViewModel> results = new ArrayList<>();
-    private Optional<Path> newPath;
+    private Optional<Path> newPath = Optional.empty();
     private int numberSucessful;
     private int totalFilesCounter;
 
@@ -54,8 +54,7 @@ public class CopyFilesTask extends Task<List<CopyFilesResultItemViewModel>> {
     }
 
     @Override
-    protected List<CopyFilesResultItemViewModel> call()
-            throws InterruptedException, IOException {
+    protected List<CopyFilesResultItemViewModel> call() throws InterruptedException, IOException {
 
         updateMessage(Localization.lang("Copying files..."));
         updateProgress(0, totalFilesCount);
@@ -104,9 +103,6 @@ public class CopyFilesTask extends Task<List<CopyFilesResultItemViewModel>> {
             }
             updateMessage(Localization.lang("Finished copying"));
 
-            if (newPath == null) {
-                return results;
-            }
             String sucessMessage = Localization.lang("Copied %0 files of %1 sucessfully to %2", Integer.toString(numberSucessful), Integer.toString(totalFilesCounter), newPath.map(Path::getParent).map(Path::toString).orElse(""));
             updateMessage(sucessMessage);
             bw.write(sucessMessage);

--- a/src/main/java/org/jabref/gui/copyfiles/CopyFilesTask.java
+++ b/src/main/java/org/jabref/gui/copyfiles/CopyFilesTask.java
@@ -103,6 +103,10 @@ public class CopyFilesTask extends Task<List<CopyFilesResultItemViewModel>> {
                 }
             }
             updateMessage(Localization.lang("Finished copying"));
+
+            if (newPath == null) {
+                return results;
+            }
             String sucessMessage = Localization.lang("Copied %0 files of %1 sucessfully to %2", Integer.toString(numberSucessful), Integer.toString(totalFilesCounter), newPath.map(Path::getParent).map(Path::toString).orElse(""));
             updateMessage(sucessMessage);
             bw.write(sucessMessage);

